src/ipc/ipc4/handler.c

```diff
@@ -379,12 +380,54 @@ static int ipc_wait_for_compound_msg(void)
 	return ret;
 }
 
+/* In ipc3 path, host driver sends pcm_hw_param message to fw and
+ * direction is included. The direction is set to each component
+ * after pipeline is complete. There is no such ipc message for
+ * ipc4 path and direction can only be figured out from dai gateway
+ * type. Ipc4 path gets direction from dai component and set
+ * it to each component after connected pipeline are complete.
+ */
+static int update_dir_to_pipeline_component(uint32_t *ppl_id, uint32_t count)
+{
+	struct ipc_comp_dev *icd;
+	struct comp_dev *dai;
+	struct list_item *clist;
+	struct ipc *ipc;
+	uint32_t latency, i;
+
+	ipc = ipc_get();
+
+	/* only dai has direction based on gateway type */
+	dai = pipeline_get_dai_comp(ppl_id[0], &latency);
+	if (!dai) {
+		tr_err(&ipc_tr, "can't find dai comp");
+		return -EINVAL;
+	}
+
+	/* set dai direction to the component in the pipline array */
+	list_for_item(clist, &ipc->comp_list) {
+		icd = container_of(clist, struct ipc_comp_dev, list);
+		if (icd->type != COMP_TYPE_COMPONENT)
+			continue;
+
+		for (i = 0; i < count; i++) {
+			if (ipc_comp_pipe_id(icd) == ppl_id[i]) {
+				icd->cd->direction = dai->direction;
+				break;
```

Comment:
How it will behave when hda copiers, both host output and input, are under the same pipeline?

Reply:
@tmleman  good point. And do we have a use case that dai copier to dai copier ?

Reply:
@RanderWang yes, mainly for validation.

Reply:
Thanks, I will redesign it